### PR TITLE
Implement styles for material theme in dxLookup

### PR DESCRIPTION
**Pull request by groshenkovamarina**

### CODE REVIEW

File: js/ui/lookup.js [hunk 1]
```diff
@@ -623,6 +629,93 @@ var Lookup = DropDownList.inherit({
...
+                    * @type_function_return boolean
+                    */
+
+                    closeOnOutsideClick: true,
+
+                    /**
+                     * @name dxLookupOptions_popupWidth
+                     * @publicName popupWidth
+                     * @type number|string|function
+                     * @default undefined @for Material
```
Comment (DokaRus) [inline js/ui/lookup.js]: But default value is `(function() { return this._$element.outerWidth(); }).bind(this)`

File: js/ui/lookup.js [hunk 2]
```diff
@@ -623,6 +629,93 @@ var Lookup = DropDownList.inherit({
...
+
+                    /**
+                     * @name dxLookupOptions_popupHeight
+                     * @publicName popupHeight
+                     * @type number|string|function
+                     * @default undefined @for Material
+                     * @type_function_return number|string
+                     */
+
+                    popupHeight: (function() { return this._setPopupHeight(MATERIAL_LOOKUP_LIST_ITEMS_COUNT); }).bind(this),
```
Comment (DokaRus) [inline js/ui/lookup.js]: >_setPopupHeight

It's confusing. Function's name tells that it is a setter but in fact, it's a getter?

File: js/ui/lookup.js [hunk 3]
```diff
@@ -623,6 +629,93 @@ var Lookup = DropDownList.inherit({
...
+                     * @type_function_return number|string
+                     */
+
+                    popupWidth: (function() { return this._$element.outerWidth(); }).bind(this),
+
+                    /**
+                     * @name dxLookupOptions_popupHeight
+                     * @publicName popupHeight
+                     * @type number|string|function
+                     * @default undefined @for Material
```
Comment (DokaRus) [inline js/ui/lookup.js]: You define a default value below

File: js/ui/lookup.js [hunk 4]
```diff
@@ -751,15 +853,56 @@ var Lookup = DropDownList.inherit({
...
+            listItems = this._list.option("items"),
+            itemsCount = listItems.length;
+
+        if(this._list.option("grouped")) {
+            this._list.scrollToItem({ group: itemsCount - 1, item: listItems[itemsCount - 1].items.length - 1 });
+        } else {
+            this._list.scrollToItem(itemsCount - 1);
+        }
+
+        this._list.scrollToItem(selectedIndex);
```
Comment (babich-a) [inline js/ui/lookup.js]: Why do we need to scroll to the last item first and then to the selectedItem?

File: js/ui/lookup.js [hunk 5]
```diff
@@ -623,6 +629,93 @@ var Lookup = DropDownList.inherit({
...
+                    * @default false @for Material
+                    */
+
+                    showPopupTitle: false,
+
+                    /**
+                    * @name dxLookupOptions_position
+                    * @publicName position
+                    * @type positionConfig
+                    * @default undefined @for Material
```
Comment (DokaRus) [inline js/ui/lookup.js]: You have described `position` config below

**Review by DokaRus — CHANGES_REQUESTED** [review]

File: testing/tests/DevExpress.ui/defaultOptions.tests.js [hunk 6]
```diff
@@ -611,6 +611,33 @@ testComponentDefaults(Lookup,
     }
 );
 
+testComponentDefaults(Lookup,
+    {},
+    {
+        usePopover: false,
+
```
Comment (babich-a) [inline testing/tests/DevExpress.ui/defaultOptions.tests.js]: Let's remove empty lines after each option.

File: js/ui/lookup.js [hunk 7]
```diff
@@ -804,6 +947,14 @@ var Lookup = DropDownList.inherit({
         this.option("focusStateEnabled") && this.focus();
     },
 
+    _popupHiddenHandler: function() {
+        this.callBase();
+
+        if(this.option("_scrollToSelectedItemEnabled")) {
+            translator.resetPosition(this._popup._$content);
```
Comment (mrjohn0011) [inline js/ui/lookup.js]: Use this._popup.$content() method

File: js/ui/lookup.js [hunk 8]
```diff
@@ -751,15 +853,56 @@ var Lookup = DropDownList.inherit({
...
+            var differenceOfOffsets = selectedListItem.offset().top - this._popup._$content.offset().top;
+
+            if(this._$element.offset().top > differenceOfOffsets) {
+                popupOffsetY = differenceOfOffsets + differenceOfHeights;
+            } else {
+                popupOffsetY = differenceOfHeights;
+                this._scrollToSelectedItem();
+            }
+
+            this._popup._changePosition({ left: 0, top: -popupOffsetY });
```
Comment (mrjohn0011) [inline js/ui/lookup.js]: Can you use this._popup.option("position") here? Don't use private methods

File: js/ui/lookup.js [hunk 9]
```diff
@@ -751,15 +853,56 @@ var Lookup = DropDownList.inherit({
...
+
+    _setPopupPosition: function() {
+        var selectedIndex = this._list.option("selectedIndex"),
+            selectedListItem = this._list._$element.find("." + this._list._selectedItemClass()),
+            differenceOfHeights = (selectedListItem.height() - this._$element.height()) / 2,
+            popupOffsetY = 0;
+
+
+        if(selectedIndex !== -1) {
+            var differenceOfOffsets = selectedListItem.offset().top - this._popup._$content.offset().top;
```
Comment (mrjohn0011) [inline js/ui/lookup.js]: use popup.$content()

File: js/ui/lookup.js [hunk 10]
```diff
@@ -751,15 +853,56 @@ var Lookup = DropDownList.inherit({
...
+        } else {
+            this._list.scrollToItem(itemsCount - 1);
+        }
+
+        this._list.scrollToItem(selectedIndex);
+    },
+
+    _setPopupPosition: function() {
+        var selectedIndex = this._list.option("selectedIndex"),
+            selectedListItem = this._list._$element.find("." + this._list._selectedItemClass()),
```
Comment (mrjohn0011) [inline js/ui/lookup.js]: Don't use double private methods anywhere. `_list._method` should be never used. Use another way to get selected item element from the list

File: js/ui/lookup.js [hunk 11]
```diff
@@ -751,15 +853,56 @@ var Lookup = DropDownList.inherit({
...
+    },
 
-        this._attachSearchChildProcessor();
+    _scrollToSelectedItem: function() {
+        var selectedIndex = this._list.option("selectedIndex"),
+            listItems = this._list.option("items"),
+            itemsCount = listItems.length;
+
+        if(this._list.option("grouped")) {
+            this._list.scrollToItem({ group: itemsCount - 1, item: listItems[itemsCount - 1].items.length - 1 });
```
Comment (mrjohn0011) [inline js/ui/lookup.js]: Check that the widget with grouped strategy does not crash if there are no items in it. Write a test

File: testing/tests/DevExpress.ui.widgets.editors/lookup.tests.js [hunk 12]
```diff
@@ -2993,6 +2994,70 @@ QUnit.module("default options", {
...
+
+    var $lookup = $("<div>").prependTo("body");
+
+    try {
+
+        var lookup = $lookup.dxLookup({ dataSource: ["blue", "orange", "lime", "purple"] }).dxLookup("instance");
+
+        assert.equal(lookup.option("popupWidth")(), $lookup.outerWidth(), "popup width match with lookup field width");
+        assert.equal(lookup.option("position").of, lookup._$element, "popup position of lookup field");
+
```
Comment (mrjohn0011) [inline testing/tests/DevExpress.ui.widgets.editors/lookup.tests.js]: remove one empty string

File: testing/tests/DevExpress.ui.widgets.editors/lookup.tests.js [hunk 13]
```diff
@@ -2993,6 +2994,70 @@ QUnit.module("default options", {
...
+        $(lookup._$field).trigger("dxclick");
+
+        assert.equal(lookup.option("popupHeight")(), $(".dx-list-item").height() * 4, "popup height contains 4 list items");
+
+        lookup.close();
+        lookup.option("popupWidth", 200);
+        lookup.option("popupHeight", 300);
+
+        $(lookup._$field).trigger("dxclick");
```
Comment (mrjohn0011) [inline testing/tests/DevExpress.ui.widgets.editors/lookup.tests.js]: Don't use private methods outside of the owner widget

File: testing/tests/DevExpress.ui.widgets.editors/lookup.tests.js [hunk 14]
```diff
@@ -2993,6 +2994,70 @@ QUnit.module("default options", {
...
+    var origIsMaterial = themes.isMaterial;
+    themes.isMaterial = function() { return true; };
+
+    var $lookup = $("<div>").prependTo("body");
+
+    try {
+
+        var lookup = $lookup.dxLookup({ dataSource: ["blue", "orange", "lime", "purple"], value: "blue" }).dxLookup("instance");
+
+        $(lookup._$field).trigger("dxclick");
```
Comment (mrjohn0011) [inline testing/tests/DevExpress.ui.widgets.editors/lookup.tests.js]: Don't use private methods outside of the owner widget

File: testing/tests/DevExpress.ui.widgets.editors/lookup.tests.js [hunk 15]
```diff
@@ -2993,6 +2994,70 @@ QUnit.module("default options", {
...
+
+        var $popup = $(".dx-popup-wrapper");
+
+        assert.roughEqual($popup.find(".dx-overlay-content").position().top, -3.5, 1, "offset of the lookup if first item is selected");
+
+        lookup.close();
+
+        lookup.option("value", "purple");
+
+        $(lookup._$field).trigger("dxclick");
```
Comment (mrjohn0011) [inline testing/tests/DevExpress.ui.widgets.editors/lookup.tests.js]: Don't use private methods outside of the owner widget

**Review by mrjohn0011 — DISMISSED** [review]

File: js/ui/lookup.js (hunk 8 above)
Comment (groshenkovamarina) [reply js/ui/lookup.js]: Can't, doesn't work, if set this._popup.option("position").offset here

File: js/ui/lookup.js [hunk 16]
```diff
@@ -358,6 +362,7 @@ var Lookup = DropDownList.inherit({
             * @type boolean
             * @default false
             */
+
```
Comment (DokaRus) [inline js/ui/lookup.js]: Remove empty line, pls

File: js/ui/lookup.js [hunk 17]
```diff
@@ -623,6 +630,78 @@ var Lookup = DropDownList.inherit({
...
+                },
+                options: {
+
+                    /**
+                    * @name dxLookupOptions_usePopover
+                    * @publicName usePopover
+                    * @type boolean
+                    * @default false @for Material
+                    */
+
```
Comment (DokaRus) [inline js/ui/lookup.js]: Remove empty line between doccoment and option for all cases below

File: js/ui/lookup.js [hunk 18]
```diff
@@ -804,6 +934,14 @@ var Lookup = DropDownList.inherit({
         this.option("focusStateEnabled") && this.focus();
     },
 
+    _popupHiddenHandler: function() {
+        this.callBase();
+
+        if(this.option("_scrollToSelectedItemEnabled")) {
+            translator.resetPosition(this._popup.$element().find(".dx-overlay-content"));
```
Comment (DokaRus) [inline js/ui/lookup.js]: Use `element()` instead of `$element()`

File: js/ui/lookup.js [hunk 19]
```diff
@@ -825,6 +963,7 @@ var Lookup = DropDownList.inherit({
         delete result.animation;
         delete result.position;
 
+
```
Comment (DokaRus) [inline js/ui/lookup.js]: Remove a redundant empty line

**Review by DokaRus — CHANGES_REQUESTED** [review]

**Comment by lukyanovas** [comment]
Separators in doc-comments have been changed in #3896 Please, rebase your PR.

File: js/ui/lookup.js [hunk 20]
```diff
@@ -623,6 +629,71 @@ var Lookup = DropDownList.inherit({
...
                 }
+            },
+            {
+                device: function() {
+                    return themes.isMaterial();
+                },
+                options: {
+
+                    /**
+                    * @name dxLookupOptions_usePopover
```
Comment (lukyanovas) [inline js/ui/lookup.js]: dxLookupOptions_usePopover => dxLookupOptions.usePopover

Comment (groshenkovamarina) [reply js/ui/lookup.js]: Fixed

**Review by DokaRus — DISMISSED** [review]

File: js/ui/lookup.js (hunk 8 above)
Comment (babich-a) [reply js/ui/lookup.js]: fixed

File: js/ui/lookup.js [hunk 21]
```diff
@@ -693,6 +767,15 @@ var Lookup = DropDownList.inherit({
         this._inkRipple = inkRipple.render();
     },
 
+    _toggleOpenState: function() {
+
```
Comment (DokaRus) [inline js/ui/lookup.js]: Remove empty line

File: js/ui/lookup.js [hunk 22]
```diff
@@ -511,6 +518,8 @@ var Lookup = DropDownList.inherit({
             * @hidden false
             * @inheritdoc
             */
+
```
Comment (DokaRus) [inline js/ui/lookup.js:520]: Remove empty line

**Review by DokaRus — APPROVED** [review]

**Pull request merged**